### PR TITLE
CI Runners for parallel cluster

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Adding some runners managers that dispatch to  AWS pcluster AMIs using the AWS autoscaler.

After reading the docs a few more times, I think it is sufficient to use the GitLab runner docker containers we have with the `docker+machine` executor. My understanding is this executor runs the runner manger in docker, and then then pipeline is launched on the "machine" using the AMI directly. This should mean that we _don't_ have to build a custom AMI with Gitlab components installed. This PR should be enough to start testing.

https://docs.gitlab.com/runner/configuration/runner_autoscale_aws/

### CODE REVIEW

**Comment by AlmightyYakob** [comment]
Also, I've got no idea how auth is supposed to work in this case. Since we're not running a job _on_ the cluster _from_ the cluster, and are instead directly attempting to spin up nodes and run jobs on EC2 images, we'll need to set up some IAM roles that allow this to be done. Karpenter is setup with roles to allow it to spin up nodes, but we're not using that here, so we'll need something similar.

File: k8s/runners/protected/pcluster/release.yml
Comment (kwryankrattiger) [inline k8s/runners/protected/pcluster/release.yml:63]: Need to figure out what this needs to be

Comment (kwryankrattiger) [reply k8s/runners/protected/pcluster/release.yml:63]: There is also a field for key-pairs. I think both of these can be expressed as environment variables, is it possible to  populate those via secrets here?

@AlmightyYakob

**Comment by kwryankrattiger** [comment]
Setup the runners and provisioner for staging in a fixup commit, this commit should be reverted before merging.

File: terraform/modules/spack/karpenter.tf
Comment (kwryankrattiger) [reply terraform/modules/spack/karpenter.tf:130]: ah nice, trying it now

File: k8s/production/karpenter/provisioners/runners/pcluster/graviton/3/provisioners.yaml
Comment (mvandenburgh) [inline k8s/production/karpenter/provisioners/runners/pcluster/graviton/3/provisioners.yaml:56]:
```suggestion
      values:
      - "us-west-2a"
      - "us-west-2b"
      - "us-west-2c"
      - "us-west-2d"
```

File: k8s/production/runners/protected/graviton/3/release.yaml
Comment (mvandenburgh) [inline k8s/production/runners/protected/graviton/3/release.yaml:51]: Why this change? Does the "512m" string not work?

**Review by mvandenburgh — COMMENTED** [review]
I left one last question, but besides that I think this is good to go.

File: terraform/modules/spack/karpenter.tf
Comment (mvandenburgh) [inline terraform/modules/spack/karpenter.tf:111]: Note, these are custom "EKS-optimized" AMIs based on the AWS parallel cluster AMI. The packer code I used to build them is on this branch - https://github.com/mvandenburgh/amazon-eks-ami/tree/pcluster-ami

Comment (mvandenburgh) [reply terraform/modules/spack/karpenter.tf:111]: If these pcluster runners are going to be a longer term thing, we may want to run this packer build in CI. The AMI images are dependent on kubernetes version, so left un-automated this is another thing we'll have to manually update when we get to the kubernetes upgrade.

**Comment by kwryankrattiger** [comment]
I am okay with this as is, but I can't merge it.

File: k8s/production/runners/protected/graviton/3/release.yaml
Comment (mvandenburgh) [reply k8s/production/runners/protected/graviton/3/release.yaml:51]: The rest of the runners in `k8s/production/runners/*` have this as a string, so I think before merging we should either back out this change, or update the rest of the runner memory requests to also be an int. Maybe the former is better, as the gitlab docs usually have these types of fields set as strings? (see `helper_cpu_limit` in this example - https://gitlab.com/gitlab-org/gitlab-runner/blob/ffdb32cb9cd1e325b5018cea2ddd6084b6ec741e/docs/executors/kubernetes.md#define-keywords-in-the-config-toml)

**Review by AlmightyYakob — APPROVED** [review]

File: k8s/production/runners/protected/pcluster/graviton/3/release.yaml
Comment (scottwittenburg) [inline k8s/production/runners/protected/pcluster/graviton/3/release.yaml:111]: I wonder if this is overwriting the `spack-intermediate-ci-signing-key` stuff mounted in the same location above?  See this job, where there don't seem to be any keys there:

https://gitlab.spack.io/spack/spack/-/jobs/7122837#L404

Comment (kwryankrattiger) [reply k8s/production/runners/protected/pcluster/graviton/3/release.yaml:111]: Delete this, use the mounted secret > the host path. Host path is the AMI location, which doesn't have secrets.

**Pull request merged**
